---
author: owent
categories:
  - Article
  - Blablabla
date: 2022-07-09 23:43:45
draft: true
id: 2206
tags: 
  - libcopp
  - coroutine
  - 协程
  - C++20
title: libcopp对C++20协程的接入和接口设计
type: post
---

## 前言

最近开的坑有点多。有点忙不过来了所以好久没写Blog了。这个C++20的协程接入一直在改造计划中，但是一直没抽出时间来正式实施。
在之前，我写过一个初版的C++20协程接入 [《libcopp接入C++20 Coroutine和一些过渡期的设计》][2] 。当时主要是考虑到 Rust也有和C++类似的历史包袱问题，所以参考了一些Rust协程改造过程中的设计。
但是后来尝试在项目中使用的时候发现还是有一些问题。首先C++20的协程并不是零开销抽象，所以强行用Rust的模式反而带来了一定开销和理解上的难度。其次原先的设计中 generator 是按类型去实现外部接入的。但是实际接入SDK的过程中我们有相当一部分类型相同但是接入流程不同的情况，再加上现在各大编译器也都已经让C++20协程的特性脱离 experimental 阶段了，有一些细节有所变化。所以干脆根据我们实际的使用场景，重新设计了下组织结构。

我们整个服务器框架的和协程相关的改造分为两部分，第一部分是 [libcopp][1] 这个库的底层支持，另一个就是现有业务流程的改造过程。我先完成了 [libcopp][1] 对C++20协程适配支持，这也是本篇分享的主要内容。

## 使用场景分析

我们来看一下协程的使用场景。首先从使用者的角色上，主要是有3类。

+ 业务层: 由于有栈协程比如是侵入式的，那么业务调用的时候其实上下文关系是需要层层透传的。
+ 调度层: 在框架层，我们要能总控和调度资源，控制生命周期。
+ 接入层: 一般用于各类SDK接入、event loop接入或者其他事件机制接入。用来把各类平台服务或者功能库抽象成统一的协程模型。

我们在新的协程模型中，针对这三种场景设计了三个不同的工具类。在我们之前的工程实践中，采用 [libcopp][1] 的有栈协程也是有这些角色分工的，做一个简单的映射关系就是:

| 功能点            | `cotask::task<T>`     | `C++20 coroutine`                | 附注                                                                                                       |
| ----------------- | --------------------- | -------------------------------- | ---------------------------------------------------------------------------------------------------------- |
| 调度层 - 任务管理 | `cotask::task<T>`     | `task_future<TRETURN, TPRIVATE>` | 前者的私有数据分配在栈上，通过接口获取。后者的私有数据就是 `TPRIVATE`                                      |
| 接入层 - 延时操作 | `task::resume(void*)` | `generator_future<TRETURN>`      | 前者通过参数 `void*` 指向自定义类型，需要外部额外检查数据有效性，后者内置 `set_value` 接口，自动管理有效性 |
| 业务层 - 组合     | `task::yield(void*)`  | `callable_future<TRETURN>`       | 前者通过参数 `void*` 指向自定义类型来接收数据，后者通过 `co_await` 的返回                                  |
| 返回值类型        | `int`                 | `自定义`                         |                                                                                                            |

C++20协程其实是对称协程的设计，即不区分主调和被调。虽然对称协程在逻辑耦合上比非对称协程更低，且开销更低。但是在实际应用场景中，还是非对称更实用一些，导致我们使用的时候很自然地就会封装成非对称的调用方式。试想这种场景：

```cpp

callable_future<int> bar ();

callable_future<int> foo() {
  // some codes
  co_await bar(); // IO operation
  // other codes
  co_return 0;
}
```

如果 `bar()` 是一个IO操作，比如保存数据到数据库，显然我们希望 `foo()` 在 `bar();` 执行结束后继续，这里就需要建立主调和被掉关系了。而实际上，C++20协程是可以做到 `bar();` 执行结束前或者执行减速后过一段时间再回到 `co_await bar();` 执行后面的代码的。

`generator_future` 只需要维护 awaitable，不需要创建协程对象，用于尽可能轻量级地允许接入层接入到协程体系中来。 而 `callable_future<TRETURN>` 和 `task_future<TRETURN, TPRIVATE>` 的相似度比较高，都会管理协程的promise，那么我们为什么要把他们拆分成两个接口呢？这是因为 `task_future` 还附带了一些管理行为，并且主调和被掉直接没有直接联系。比如事件 `A` 可能触发一系列操作 `B`, `C` 等，并且不关心它们的结果，就可以用 `task_future` 。

对于 `callable_future` ， 我们假定它的使用场景为协程的嵌套调用，为了方便业务层使用已有的协程或者 `generator_future` 组装出满足自己业务需要的协程流程。这个时候，我们可以假定大多数情况下它是会被一个caller `co_await` ，且上下文数据不会被转移。然后可以简化内部的数据结构，降低大多数使用场景下不必要的开销。

另外还有一种使用场景是在 **调度层** 我们需要控制协程任务的生命周期，如果它超时或者被 `kill` 了。都需要层层传递通知内部正在等待的协程强行终止，这样即便内部wait for ever了，我们仍然能够强制释放资源，而不会发生泄露。这里我们需要对协程的主调、被调关系进行管理。同时一个协程还可能被多个其他协程所 `co_await` ，resume行为可能由用户或者接入层触发，也可能由析构或者调度层管理的时候触发，所以我们还需要管理好生命周期。

## C++20协程接入的设计模型

除了上述使用场景以外，不同编译器对C++20协程的实现也有些差异。虽然都是符合标准的，但是某些接口的调用先后顺序和析构（特别是临时对象）的先后顺序是有差异的。显然我们 [libcopp][1] 一个非常重要的要求就是抹平这些差异。以此为目的，我们设计的几个组件的主要流程和调用关系如下:


| 模块                               | `callable_future` - 业务层                                  | `task_future` - 调度层（框架）                            | `generator_future` - 接入层（扩展）   |     |
| ---------------------------------- | ----------------------------------------------------------- | --------------------------------------------------------- | ------------------------------------- | --- |
| `promise_type`                     | `data:waiting,callers,status,DATA`                          | `data:waiting,callers,status,context`                     | -                                     |     |
| `promise_type:unhandled_exception` | 不捕获                                                      | `start/resume` 接口后 `rethrow`                           | -                                     |     |
| `promise_type:生命周期`            | 托管到 `future`                                             | 关联到 `context` , 析构时取消关联                         | -                                     |     |
| `promise_type:Execution`           | 首次被 `co_await` 时开始执行,未被 `co_await` 则视为放弃执行 | 首次调用 `start/resume` 接口后开始执行                    | -                                     |     |
| `promise_type:final_suspend`       | 所有 `callers` 调用 `resume()`                              | 所有 `callers` 调用 `resume()`                            | -                                     |     |
|                                    |                                                             |                                                           |                                       |     |
| `context`                          | -                                                           | `data:task_id,handle,unhandle_exception`                  | `data:caller`                         |     |
| `context:析构/set_value`           | -                                                           | `handle.promise().waiting.resume()`                       | `caller.resume()`                     |     |
|                                    | -                                                           | `handle.promise().resume()`                               |                                       |     |
|                                    | -                                                           | `handle.promise().destroy()`                              |                                       |     |
|                                    |                                                             |                                                           |                                       |     |
| `future`                           | `data:handle`                                               | `data:context`                                            | `data:context`                        |     |
| `future:析构`                      | `handle.promise().resume()`                                 |                                                           |                                       |     |
|                                    | `handle.promise().destroy()`                                |                                                           |                                       |     |
|                                    |                                                             |                                                           |                                       |     |
| `awaitable`                        | `data:caller,callee`                                        | `data:caller,context(raw ptr)`                            | `data:caller,context(raw ptr)`        |     |
| `awaitable:析构`                   | 由 `promise_type:final_suspend` 保证有效性,无需额外处理     | `detach()`                                                | `detach()`                            |     |
|                                    |                                                             |                                                           |                                       |     |
| `awaitable:detach`                 | `callee.promise().remove_caller(caller)`                    | `context->handle.promise().remove_caller(caller)`         | `context->remove_caller(caller)`      |     |
|                                    | `caller.waiting = nullptr`                                  | `caller.waiting = nullptr`                                | `context->caller = nullptr`           |     |
|                                    | 故障处理:状态传递( `callee.status = caller.status` )        | 故障处理(exiting):依据 `context->handle.promise().status` |                                       |     |
|                                    |                                                             | 故障处理:依据 `caller.status`                             |                                       |     |
|                                    |                                                             |                                                           |                                       |     |
| `awaitable:await_suspend`          | `caller.promise().waiting = callee`                         | `caller.promise().waiting = callee`                       | `this->caller = caller`               |     |
|                                    | `callee.promise().add_caller(caller)`                       | `context->handle.promise().add_caller(caller)`            | `context->add_caller(caller)`         |     |
|                                    |                                                             |                                                           |                                       |     |
| `awaitable:await_resume`           | `detach()`                                                  | `detach()`                                                | `detach()`                            |     |
|                                    | `callee.promise().waiting.resume()`                         | 转移数据(返回 `context->handle.promise().status`)         | 转移数据(ready/依据 `caller.status` ) |     |
|                                    | 转移数据(强制进入ready/依据 `caller.status`)                |                                                           |                                       |     |
|                                    |                                                             |                                                           |                                       |     |

除了这些以外，和 [libcopp][1] 内的有栈协程一样，我们提供了一些特殊的错误流程， **超时** 、 **取消** 和 **Killed** 。所以，如果 `TRETURN` 不是整数类型，我们就要求用户要提供 `copp::promise_error_transform<TRETURN>` 的特化，接受 `promise_status` 类型，返回 `TRETURN` 类型或者可以隐式转换到 `TRETURN` 的类型。

这几个组件其实是有一部分公共的部分的，对于这部分我们可以抽离出来，另外再结合前面提到的生命周期管理的一些要点，我简单总结了一下如下:

+ 公共逻辑
  + `promise_error_transform<T>::operator()(promise_status) const noexcept` : 故障转换 `cancle/timeout/kill`
  + `promise_type` 的公共数据
    + 正在等待的对象(用于传递 `cancle/timeout/kill` ): `waiting：coroutine_handle<std_coroutine_promise_base>`
    + 状态: `status：created/running/done/cancle/kill/timeout`
  + `awaitable` 的公共数据
    + **生命周期从属于调用方的 `promise_type`，且在 `co_await` 完成前被析构，此时被等待方的handle可能已经失效**
      + 析构时: `caller` 有效，`callee` 可能无效
      + `await_ready/await_suspend` 时: `caller` 有效，`callee/generator_context` 有效
        > **注意：在多线程模式下，`await_suspend` 由发起 `co_await` 的线程在挂起协程后执行，如果这时候handle被传递到其他线程且被resume了。**
        > **这时候当前的awaitable对象会被析构，且 `await_suspend` 未执行完，这里需要额外的同步机制或者在 `await_suspend` 中传出handle后不再使用成员。**
        > *See [Coroutines (C++20) - cppreference.com](https://en.cppreference.com/w/cpp/language/coroutines#co_await)*
      + `await_resume` 时: caller有效，`callee/generator` 可能无效
        + **通过callee在 `final_suspend` 时强制触发caller的resume，来保证callee此时有效**
        + **generator的 `context` 在析构时，也需要强制触发resume，来保证 `context` 此时有效**
        + 但是第一次调用调用 `resume()` 后仍然可能导致 `callee/generator_context` 被释放，此时需要 `await_suspend` 配置
    + 调用者: `caller：promise_base_type:handle_delegate`
  + 通过 `callable_promise/callable_context` 链式传递关系
    + out of control: resume后需要检查状态，如果被kill，需要尽可能执行到结束
- [x] `generator_future` （IO操作）绑定类型
  - [x] 生命周期安全（调度层发起和恢复，内部调用）
    - [x] `context` 析构前要把关联的 `handle` 直接 `resume` 掉（可能转为cancle状态）
  - [x] `context:set_value`
  - [x] 事件转换（事件回调）
    - [x] `await_suspend_callback(shared_ptr<generator_context>)` : 允许续期 context
    - [x] `await_resume_callback(const generator_context&)` : context 可能正在析构，不允许续期
- [x] `task_future`
  - [x] 生命周期安全（业务层发起，调度层发起，框架层管理，内部调用）
    - [x] 单向引用关系:
      - [x] `task_future` 对 `context` 强引用
      - [x] `promise_type` 对 `context` 强引用,但是最后一个 `task_future` 析构时要强制kill掉
      - [x] `context` 对 `promise_type` 弱引用(通过 `handle_delegate` )
    - [x] `context` 析构(最后一个 `task_future` 释放)前要把关联的 `handle` 一直 `resume` 到 `done` 为止
  - [x] `await/then`
    - ~~循环等待检查~~
  - [x] `kill`
  - [x] 任务管理: `task_manager`
    - [x] 超时管理
- [x] `callable_future`
  - [x] 可移动，不可复制（仅内部调用）
  - [x] 管理handle生命周期(否则awaitable和promise的释放顺序是UB)
  - [x] 传递失败状态
- [x] 特殊的`callable_future`
  - [x] any( `callable_future/generator_future/task_future` )
  - [x] all( `callable_future/generator_future/task_future` )
  - [x] some( `callable_future/generator_future/task_future` )
- Test
  - [x] task->callable->generator(强制task超时恢复要能正确处理内部两个awaitable)
    - 此时强制执行 `task.coroutine_handle.resume()` callable能够收到 `await_resume` , 但 `generator` 无法获知上级状态变化
  - [x] promise的析构和awaitable的 `await_resume` 执行顺序是未定义的。所以不应该在 `await_resume` 里尝试获取promise
    - [x] 或者在 `final_suspend():await_ready()` 返回 `false` 然后在资源引用移除时执行 `handle.destroy()`
    - [x] `callable_future`
    - [x] `task_future`
    - [x] any/all/some
      - [x] 基本流程
      - [x] kill和状态传递


## 设计模式适配 - `task_manager`, `some()`, `any()` 和 `all()`

## Benchmark

压力测试仍然是分为trivial类型返回和非trivial类型返回两部分。非trivial类型因为有构造和析构函数，是会影响编译优化的效果的。

按 https://en.cppreference.com/w/cpp/language/coroutines#Heap_allocation 的描述，某些场景下C++20协程是可以把promise的分配优化掉的。那么在高性能的接口层 （ `callable_future` ） 我们当然也想尽可能利用好编译器的特性。在 [单元测试][3] 里 `libcopp_sample_benchmark_std_couroutine_callable_recursive` 相关的项目中。可以看到 create 时间很长，但是 resume 时间为0，这估计就是触发这个优化+尾调用优化了，直接创建协程的时候就把后面的跑完了。

## 接入过程中的易踩坑点

+ 临时对象的析构
+ promise_type的类型转换

## 从有栈协程迁移

`task_manager`

## 最后

欢迎有兴趣的小伙伴交流分享。

[1]: https://github.com/owent/libcopp
[2]: https://owent.net/2020/2004.html
[3]: https://github.com/owent/libcopp/actions
